src/date.c

```diff
@@ -1,18 +1,983 @@
+                          like x = {&tp.year, &tp.month, .... } */
+  tp_arr[0] = &tp.year; tp_arr[1] = &tp.month; tp_arr[2] = &tp.day;
+  tp_arr[3] = &tp.hour; tp_arr[4] = &tp.min; tp_arr[5] = &tp.sec;
+  tp_arr[6] = &tp.msec;
+  
+  memset(&tp, 0, sizeof(tp));
+  
+  if(breaktimefunc != NULL) {
+    breaktimefunc(current, &tp);
+  }
+  
+  for(i = 0; i < ARRAY_SIZE(tp_arr); i++) {
+    if(!isnan(a->args[i])) {
+      *tp_arr[i] = (int)a->args[i];
+    }
+  }
+
+  return maketimefunc(&tp);
+}
+
+static etime_t d_time_number_from_arr(struct v7 *v7, val_t this_obj, val_t args,
+            int start_pos, fbreaktime breaktimefunc, fmaketime makefilefunc) {
+  etime_t ret_time = INVALID_TIME;
+  long cargs;
+
+  val_t objtime = i_value_of(v7, this_obj);
+
+  if((cargs=v7_array_length(v7, args)) >= 1 && objtime != V7_TAG_NAN) {
+    int i;
+    struct dtimepartsarr a = {{INVALID_TIME, INVALID_TIME, INVALID_TIME,
+                    INVALID_TIME, INVALID_TIME, INVALID_TIME, INVALID_TIME}};
+    etime_t new_part = INVALID_TIME;
+    
+    for(i=0; i < cargs && (i+start_pos < tpmax); i++) {
+      new_part = i_as_num(v7, v7_array_at(v7, args, i));
+      if(isnan(new_part)) {
+        break;
+      }
+      
+      a.args[i+start_pos] = new_part;
+    }
+    
+    if(!isnan(new_part)) {
+      etime_t current_time = v7_to_double(objtime);
+      ret_time = d_changepartoftime(&current_time, &a,
+                                breaktimefunc, makefilefunc);
+    }
+  }
+  
+  return ret_time;
+}
+
+
+static val_t d_setTimePart(struct v7 *v7, val_t this_obj, val_t args,
+            int start_pos, fbreaktime breaktimefunc, fmaketime makefilefunc) {
+  val_t n;
+  etime_t ret_time = d_time_number_from_arr(v7, this_obj, args, start_pos,
+                                                breaktimefunc, makefilefunc);
+                                                                       
+  n = v7_create_number(ret_time);
+  v7_set_property(v7, this_obj, "", 0, V7_PROPERTY_HIDDEN, n);
+                                                                       
+  return n;
+}
+
+/*++++++++++++++ API +++++++++++++++*/
+
+/* constructor */
+static val_t Date_ctor(struct v7 *v7, val_t this_obj, val_t args) {
+  etime_t ret_time = INVALID_TIME;
+  if(v7_is_object(this_obj) && this_obj != v7->global_object) {
+    long cargs = v7_array_length(v7, args);
+    if(cargs <=0 ) {
+      /* no parameters - return current date & time */
+      d_gettime(&ret_time);
+    } else if(cargs == 1) {
+      /* one parameter */
+      val_t arg = v7_array_at(v7, args, 0);
+      if(v7_is_string(arg)){ /* it could be string */
+        size_t str_size;
+        const char* str = v7_to_string(v7, &arg, &str_size);
+        if(!d_timeFromString(&ret_time, str, str_size)) {
+          /* is this string is not valid date string,
+           trying to convert it to number */
+          char* endptr;
+          ret_time = strtoll(str, &endptr, 10);
```

Comment:
Could we use `i_as_num` here as well?

I noticed that v8 doesn't accept `new Date("1.0")` while if we use i_as_num here it would convert it to `1`; but I don't know if the ecma standard regulates that.